src/NumericalAlgorithms/LinearOperators/PartialDerivatives.cpp

```diff
@@ -0,0 +1,186 @@
+               1.0,                      // overall multiplier
+               matrix.data(),            // matrix
+               matrix.spacing(),         // rows of matrix including padding
+               input,                    // input
+               matrix.columns(),         // rows of input
+               0.0,                      // overwrite output with result
+               result,                   // result
+               matrix.rows());           // rows of result
+}
+}  // namespace
+
+template <typename SymmList, typename IndexList, size_t Dim>
+void logical_partial_derivative(
+    const gsl::not_null<TensorMetafunctions::prepend_spatial_index<
+        Tensor<DataVector, SymmList, IndexList>, Dim, UpLo::Lo,
+        Frame::Logical>*>
+        logical_derivative_of_u,
+    const gsl::not_null<gsl::span<double>*> buffer,
+    const Tensor<DataVector, SymmList, IndexList>& u,
+    const Mesh<Dim>& mesh) noexcept {
+  static_assert(
+      Dim > 0 and Dim < 4,
+      "logical_partial_derivative is only implemented for 1, 2, and 3d");
+  const size_t num_grid_points = mesh.number_of_grid_points();
+  ASSERT(buffer->size() >= num_grid_points,
+         "The buffer in logical_partial_derivative must be at least of size "
+             << num_grid_points << " but is of size " << buffer->size());
+
+  destructive_resize_components(logical_derivative_of_u,
+                                mesh.number_of_grid_points());
+  const Matrix identity{};
+  std::array<std::reference_wrapper<const Matrix>, Dim> diff_matrices{
+      make_array<Dim, std::reference_wrapper<const Matrix>>(identity)};
+  for (size_t d = 0; d < Dim; ++d) {
+    gsl::at(diff_matrices, d) =
+        std::cref(Spectral::differentiation_matrix(mesh.slice_through(d)));
+  }
+
+  // It would be possible to check if the memory is contiguous and then
+  // differentiate all components at once. Note that the buffer in that case
+  // would also need to be the size of all components.
+  for (size_t storage_index = 0; storage_index < u.size(); ++storage_index) {
+    const auto u_tensor_index = u.get_tensor_index(storage_index);
+    const auto xi_deriv_tensor_index = prepend(u_tensor_index, 0_st);
+    apply_matrix_in_first_dim(
+        logical_derivative_of_u->get(xi_deriv_tensor_index).data(),
+        u[storage_index].data(), diff_matrices[0], num_grid_points);
+    for (size_t i = 1; i < Dim; ++i) {
+      const auto deriv_tensor_index = prepend(u_tensor_index, i);
+      DataVector& deriv_component =
+          logical_derivative_of_u->get(deriv_tensor_index);
+      size_t chunk_size = diff_matrices[0].get().rows() *
+                          (i == 1 ? 1 : gsl::at(diff_matrices, 1).get().rows());
+      raw_transpose(make_not_null(deriv_component.data()),
+                    u[storage_index].data(), chunk_size,
+                    num_grid_points / chunk_size);
+      apply_matrix_in_first_dim(buffer->data(), deriv_component.data(),
```

Comment:
So I understand, could you please explain the underlying manipulations here? I'm fuzzy about the transpose here...I guess it's because the deriv matrix assumes row-first or column-first ordering? But if it's just a vector, what does transposing do?

Reply:
The transpose actually changes which coordinate varies fastest. We store the data as {xi, eta, zeta}, which means we can do the xi-derivative as a matrix-matrix multiply (very efficient). To do the eta-derivative we can either perform one matrix-vector multiply hopping through the data with a specified stride, or do a transpose to the data layout {eta, zeta, xi} and then do a matrix-matrix multiply. It turns out that (as long as you don't do a lot of extra allocations) doing a transpose followed by a matrix-matrix operation is faster. Afterwords you need to transpose the derivative back from {eta, zeta, xi} to {xi, eta, zeta}, hence the second transpose. Does that make sense?